### PR TITLE
Orca: Restore batch before hook `before_train_iter`

### PR DESCRIPTION
**Pull request by leonardozcm**
## Description

In hook `before_train_iter`, a user expects to see the batch in the original format, so we restore batch format before  `before_train_iter` is called. Also we recycle `self.batch` attribute when this iteration is over.

### CODE REVIEW

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 1]
```diff
@@ -475,20 +475,27 @@ def _train_batch(self, batch, batch_info=None, callbacks=None):
...
         # unpack features into list to support multiple inputs model
+        # and restore batch to what it should be.
         features, target = batch
+        if torch.is_tensor(features):
+            self.batch = features, target
+        elif isinstance(features, (tuple, list)):
+            self.batch = *features, target
+        else:
+            invalidInputError(False,
+                              "Features should be tensor, list/tuple, "
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:487]: tensor __or__ ..., to keep the same as the error below.

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 2]
```diff
@@ -475,20 +475,27 @@ def _train_batch(self, batch, batch_info=None, callbacks=None):
...
+                              "but got {}".format(type(features)))
+        self.call_hook(call_backs=callbacks, fn_name="before_train_iter")
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:489]: call_backs is a bit strange? change to callbacks?

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 3]
```diff
@@ -475,20 +475,27 @@ def _train_batch(self, batch, batch_info=None, callbacks=None):
                 calculate averages.
 
         """
-        # TODO: restore batch to what it should be.
-        self.call_hook(call_backs=callbacks, fn_name="before_train_iter")
         # unpack features into list to support multiple inputs model
+        # and restore batch to what it should be.
         features, target = batch
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:480]: is this line correct? Why you first split and then restore and then split (in Line493)?

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py (hunk 2 above)
Comment (leonardozcm) [reply python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:489]: sure

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py (hunk 3 above)
Comment (leonardozcm) [reply python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:480]: > is this line correct? Why you first split and then restore and then split (in Line493)?

`first split and restore` is meant to recover batch to how user defined in dataloader
the last split(with a star-pack * operation) is to ensure features after `before_train_iter` are fed into self.model (No interal operations on features)

And all actions passed indicates that this adjustment works well as before I think.

**Pull request merged**